### PR TITLE
Initial port of ABC difficulty adjustment change (UAHF REQ-7)

### PR DESCRIPTION
**Pull request by ftrader**
This is an initial port of ABC's difficulty change implemening UAHF REQ-7 .
It includes a unit test which still needs to be extended, hence this is marked as WIP.

I've tested that it builds and passes 'make check' .
The are still some failures in the qa tests I need to look at (I had fails on `bip9-softforks`, `excessive.py --extensive` & `p2p-acceptblock.py` during extended run)

### CODE REVIEW

**Comment by ftrader** [comment]
At least bip9-softforks.py and p2p-acceptblock.py are sporadic failures - I retried them and they passed.
Still need to look at excessive.py, whether that's a consistent failure and if it has any relationship with changes of this PR (I doubt it).

**Comment by sickpig** [comment]
@ftrader are those failures (both the sporadic and the one associated with `excessive.py`) related to your changes? Or this is something that applies at the current status of the `dev` branch? 

What's missing in the unit test you added?

**Comment by gandrewstone** [comment]
the problem with excessive.py is that it takes a long time to generate 1+ MB of transactions, send them to 3 other local nodes and validate them there.  So the propagation of the block is slow.  I think that its gotten slower -- we may be throttling the transaction propagation too much.  But this is more a regtest problem than a real issue since the real network does not create 1+ MB of transaction in seconds.  I am making a PR that solves this issue by reducing the number of transactions generated by excessive.py.  

This issue is not a problem with this PR

**Comment by sickpig** [comment]
@gandrewstone why not trying to artificially makes the coinbase bigger? 
see: https://github.com/Bitcoin-ABC/bitcoin-abc/commit/9c1ddd0ae24209e87f69ca30798b9f16082176ba

**Comment by ftrader** [comment]
@sickpig : the unit test needs to be extended with another case to test the 20% drop in diff (without hitting POW limit).

UPDATE: I see deadalnix has submitted an extension of this test in https://reviews.bitcoinabc.org/D353 .

**Comment by ftrader** [comment]
Updated, taking this out  of WIP now

File: src/pow.cpp [hunk 1]
```diff
@@ -20,34 +20,65 @@ unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHead
...
 
-    // Go back by what we want to be 14 days worth of blocks
-    int nHeightFirst = pindexLast->nHeight - (params.DifficultyAdjustmentInterval()-1);
-    assert(nHeightFirst >= 0);
-    const CBlockIndex* pindexFirst = pindexLast->GetAncestor(nHeightFirst);
-    assert(pindexFirst);
+    // Do not go lower than the POW limit.
+    uint32_t nBits = pindexLast->nBits;
+    if (nBits == nProofOfWorkLimit) {
+        return nProofOfWorkLimit;
```
Comment (ptschip) [inline src/pow.cpp:58]: should this be:

if (nBits <= nProofOfWorkLimit) {

Comment (deadalnix) [reply src/pow.cpp:58]: This is an just an early bail out. It's not expected that is can be lower than the minimum considering it comes from a existing block.

If you look after this, the code figure out if we match the condition under which the difficulty is lowered, lowers it is so and then check for minimum PoW again.

Comment (ftrader) [reply src/pow.cpp:58]: I would agree.

Comment (ptschip) [reply src/pow.cpp:58]: ok fair enough...

Comment (deadalnix) [reply src/pow.cpp:58]: To be 100% complete, the actual check is line 79.

**Pull request merged**